scripts/generators/stubGenerator.sh

```diff
@@ -0,0 +1,333 @@
+      if [ ${#paramType} == 0 ]; then
+        paramType=$paramName
+        paramName=tmp$i
+      fi
+
+      # handle chan param type
+      if [ "$paramType" == "chan" ] && [ ${#optParamType} != 0 ]; then
+        paramType+=" "$optParamType
+      fi
+
+      if [ "$paramName" == "chan" ]; then
+        paramType=$paramName" "$paramType
+        paramName=tmp$i
+      fi
+
+      paramNames+=("$paramName")
+      paramTypes+=("$paramType")
+    done
+  fi
+}
+
+createMethodBody() {
+  # if method is IsInterfaceNil, write special return and return
+  if [[ $methodName == *"IsInterfaceNil"* ]]; then
+    { echo "return stub == nil";
+      echo -e "}\n";
+      } >> "$stubPath"
+    return
+  fi
+
+  # add the check to stub member to not be nil
+  echo "if stub.$stubField != nil {" >> "$stubPath"
+
+  stringParamNames=""
+  getStringParamNames
+
+  # add return statement calling stub member
+  # if there is no return type, add it without return
+  # otherwise, return it with the provided params
+  if [[ ${#returnTypesArr} == 0 ]]; then
+    writeWithNoReturn
+  else
+    writeWithReturn
+  fi
+}
+
+getStringParamNames() {
+  for i in "${paramNames[@]}"; do
+    stringParamNames+=$i
+    stringParamNames+=", "
+  done
+
+  if [[ ${#stringParamNames} != 0 ]]; then
+    # remove the last ", " added with last param
+    stringParamNames=${stringParamNames::-2}
+  fi
+}
+
+writeWithNoReturn() {
+  { echo "stub.$stubField($stringParamNames)";
+    echo "}";
+    echo -e "}\n";
+    } >> "$stubPath"
+}
+
+writeWithReturn() {
+  { echo "return stub.$stubField($stringParamNames)";
+  echo "}";
+  } >> "$stubPath"
+
+  # compute default values to return when stub member is not provided, separated by comma
+  toReturn=""
+  extractDefaultReturn
+
+  # write the final return statement to file with default params and close the method
+  { echo "return $toReturn";
+    echo -e "}\n";
+    } >> "$stubPath"
+}
+
+extractDefaultReturn() {
+  for i in "${returnTypesArr[@]}"; do
+    case $i in
+      bool)
+        toReturn+="false, "
+        ;;
+      string | core.PeerID)
+        toReturn+="\"\", "
+        ;;
+      int | uint32 | uint64 | int32 | int64 | float | float32 | float64 | byte)
+        toReturn+="0, "
+        ;;
+      *)
+        toReturn+="nil, "
+        ;;
+    esac
+  done
+
+  if [[ ${#toReturn} != 0 ]]; then
+    # remove the last ", " added with last param
+    toReturn=${toReturn::-2}
+  fi
+}
+
+computeUpdatedParameters() {
+  updatedParameters+="("
+  for i in "${!paramNames[@]}"; do
+    updatedParameters+="${paramNames[$i]%%"."*}" # until first dot, in case there are variable params
+    updatedParameters+=" "
+    updatedParameters+="${paramTypes[$i]}"
+    updatedParameters+=", "
+  done
+
+  if [[ ${#updatedParameters} != 1 ]]; then
+    # remove the last ", " added with last param
+    updatedParameters=${updatedParameters::-2}
+  fi
+
+  updatedParameters+=")"
+}
+
+removeCommentsFromMethodLine() {
```

Comment:
The function definition must be placed before any calls to the function !!! (bash is not a compiler)